proposals/0724-qualified-lists.rst

```diff
@@ -0,0 +1,397 @@
+----------
+
+Problems with Type Class-driven overloading
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``OverloadedLists`` works by desugaring to a type-class-overloaded function. That leads to certain general shortcomings:
+
+* It is a module-wide setting
+
+  * Anecdotally, people would rather avoid ``OverloadedLists`` than deal with overloaded lists in the entire module.
+
+  * It's possible that this is one reason this extension isn't a default in GHC202X language editions, despite being in GHC for a long time.
+
+* Type inference ambiguity.
+
+  * Consider the following code:
+
+    ::
+
+      output :: Foldable t => t -> IO ()
+
+      main = do
+        output [False, True]
+
+        -- output $ V.update myVec [(0, True)]
+
+    This originally works with no extensions, due to the list literal being typed to concrete ``[Bool]``. But say the developer wants to call an API using ``Vector`` and use ``Vector`` literals; adding ``OverloadedLists`` would cause ambiguity to the existing location because it is now no longer concretely ``[Bool]`` but ``(IsList l, Item l ~ Bool) => l``.
+
+This proposal would allow using a module qualifier to say precisely which function to desugar to, rather than using type classes, in a similar manner as ``-XQualifiedDo``. This would allow writing the previous code as
+
+::
+
+  {-# LANGUAGE QualifiedLists #-}
+
+  main = do
+    output [False, True]
+
+    output $ V.update myVec V.[(0, True)]
+
+The existing locations would continue working as ``[Bool]``, while the new literal would unambiguously desugar to the equivalent of ``V.fromList [(0, True)]``.
+
+Inability to use heterogeneous lists
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With ``-XOverloadedLists`` we can never write the literal ``[4, "hello", True]``, because that desugars to ``fromList [4, "hello", True]`` which is ill-typed regardless of ``fromList``. That is annoyingly restrictive, because with heterogeneous lists, it's perfectly fine to write
+
+::
+
+  4 `HCons` "hello" `HCons` True `HCons` HNil :: HList [Int, String, Bool]
+
+and it would be convenient to use list literals instead. This was even explicitly listed as a restriction in the original ``OverloadedLists`` `design <https://gitlab.haskell.org/ghc/ghc/-/wikis/overloaded-lists>`_.
+
+This proposal would desugar list literals to a build-like form instead, so that ``M.[4, "hello", True]`` desugars to
+
+::
+
+  M.buildList 3 (\cons nil -> 4 `cons` ("hello" `cons` (True `cons` nil)))
+
+For a suitable ``M.buildList``, this is enough to support heterogenous list literals: see *Section 4.2 Heterogeneous Lists*.
+
+Proposed Change Specification
+-----------------------------
+
+Introduce ``-XQualifiedLists`` that desugars literal list syntax to function calls in a similar way to ``-XQualifiedDo`` (`docs <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/qualified_do.html>`_, `proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0216-qualified-do.rst>`_).
```

Comment:
add ", but with an entirely different desugaring.  It's not like `QualifiedStrings`!